File: AndroidSerializable/Parcelable/extendsapplication/src/main/java/com/haoyudu/extendsapplication/MainActivity.java
package com.haoyudu.extendsapplication;

import androidx.appcompat.app.AppCompatActivity;
import android.os.Bundle;

public class MainActivity extends AppCompatActivity {

    /**进程间通信的实现,除了通过intent相互传递还能怎样实现?
     *  --从创建实例对象开始就存在有相同的成员-->重写Application(添加成员,这样继承了Applicationi就等于共享了成员)
     * */

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
    }
}

File: KotlinTest/KotlinAndroid/app/src/main/java/com/haoyudu/kotlinandroid/MainActivity.kt
package com.haoyudu.kotlinandroid

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        val viewModel=ViewModelProvider.AndroidViewModelFactory(this.application).create(MyViewModel::class.java)
        viewModel.number.observe(this, Observer { num_text.text=it.toString() })
        btn_add.setOnClickListener{
            viewModel.modifyNumber(1)
        }
        btn_subtract.setOnClickListener{
            viewModel.modifyNumber(-1)
        }
    }
}